mm2src/coins/nft_storage/sql_storage.rs

```diff
@@ -0,0 +1,885 @@
+                .sql_builder()
+                .and_where(format!("block_timestamp <= '{}'", filters.to_date.unwrap()));
+        }
+    }
+    drop_mutability!(sql_builder);
+    Ok(sql_builder)
+}
+
+fn nft_table_builder_preimage<'a>(conn: &'a Connection, table_name: &'a str) -> MmResult<SqlQuery<'a>, SqlError> {
+    let sql_builder = SqlQuery::select_from(conn, table_name)?;
+    Ok(sql_builder)
+}
+
+fn finalize_nft_list_sql_builder(sql_builder: &mut SqlQuery, offset: usize, limit: usize) -> MmResult<(), SqlError> {
+    sql_builder.field("nft_list.details_json")?.offset(offset).limit(limit);
+    Ok(())
+}
+
+fn finalize_nft_history_sql_builder(sql_builder: &mut SqlQuery, offset: usize, limit: usize) -> MmResult<(), SqlError> {
+    sql_builder
+        .field("nft_history.details_json")?
+        .offset(offset)
+        .limit(limit);
+    Ok(())
+}
+
+fn nft_from_row(row: &Row<'_>) -> Result<Nft, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn tx_history_from_row(row: &Row<'_>) -> Result<NftTransferHistory, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn token_address_id_from_row(row: &Row<'_>) -> Result<NftTokenAddrId, SqlError> {
+    let token_address: String = row.get("token_address")?;
+    let token_id_str: String = row.get("token_id")?;
+    let token_id = BigDecimal::from_str(&token_id_str).map_err(|_| SqlError::from(FromSqlError::InvalidType))?;
+    Ok(NftTokenAddrId {
+        token_address,
+        token_id,
+    })
+}
+
+fn insert_nft_in_list_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_list_table_name(chain);
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            token_address, token_id, chain, amount, block_number, contract_type, details_json
+        ) VALUES (
+            ?1, ?2, ?3, ?4, ?5, ?6, ?7
+        );",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn insert_tx_in_history_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_tx_history_table_name(chain);
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            transaction_hash, chain, block_number, block_timestamp, contract_type,
+            token_address, token_id, status, amount, collection_name, image, token_name, details_json
+        ) VALUES (
+            ?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13
+        );",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn insert_chain_into_scanned_blocks_sql() -> MmResult<String, SqlError> {
+    let table_name = scanned_nft_blocks_table_name();
+    validate_table_name(&table_name)?;
+    Ok(format!("INSERT OR IGNORE INTO {} (chain) VALUES (?1);", table_name))
+}
```

Comment:
I believe this can be merged with `update_last_scanned_block_sql` by using `INSERT OR REPLACE`

Reply:
hmm yeah, I see, in my code I use `insert` function and right after this I call `update`. Its better just to have `INSERT OR REPLACE`.
Created `upsert_last_scanned_block_sql` function.